performance/multi-node/random_access.c

```diff
@@ -0,0 +1,1081 @@
+	 FI_WRITE | FI_SEND | FI_RECV | FI_REMOTE_READ | FI_REMOTE_WRITE)
+
+/* Program-level defines */
+#define MAX_ALIGNMENT 65536
+/* #define MAX_MSG_SIZE (1<<22) */
+#define MAX_MSG_SIZE (8*1024) /* Current GNI provider max send size */
+#define MYBUFSIZE (MAX_MSG_SIZE + MAX_ALIGNMENT)
+
+#define TEST_DESC "Libfabric Random Access Test"
+#define HEADER "# " TEST_DESC " \n"
+#ifndef FIELD_WIDTH
+#   define FIELD_WIDTH 20
+#endif
+#ifndef FLOAT_PRECISION
+#   define FLOAT_PRECISION 2
+#endif
+
+/* Random number generator */
+#define POLY 0x0000000000000007UL
+#define PERIOD 1317624576693539401L
+
+/* Define 64-bit types and corresponding format strings
+ * for printf() and constants
+ */
+#define FSTR64 "%ld"
+#define FSTRU64 "%lu"
+#define ZERO64B 0LL
+#define HEAD_RANK (rank == 0)
+#define UNLOCKED 0
+#define LOCKED(rank) (rank + 1)
+
+#define CTPM_SIMPLE_TYPE_FUNC(LABEL, type, op) \
+		static type ctpm_##LABEL (void *src, int total_ranks) \
+		 \
+		{ \
+			type *vals; \
+			int i; \
+			type ret; \
+			\
+			vals = calloc(total_ranks, sizeof(type)); \
+			assert(vals); \
+			\
+			ctpm_Allgather(src, sizeof(type), vals); \
+			\
+			ret = vals[0]; \
+			for (i = 1; i < total_ranks; i++) \
+				ret = ret op vals[i]; \
+			\
+			ctpm_Barrier(); \
+			\
+			free(vals); \
+			\
+			return ret; \
+		}
+
+#define SCRATCH_SIZE 8
+
+/* global variables */
+double gups;
+double errors_fraction;
+double check_time;
+int failure;
+uint32_t scratch[SCRATCH_SIZE];
+
+/* Allocate main table (in global memory) */
+uint64_t *hpcc_table;
+uint64_t *hpcc_lock;
+
+static uint64_t global_start_my_proc;
+
+static int rx_depth = 512;
+static int thread_safe = 1; // default
+
+struct peer_data {
+	void *r_table_addr;
+	uint64_t r_table_key;
+	void *r_lock_addr;
+	uint64_t r_lock_key;
+};
+
+struct peer {
+	struct fid_ep *ep;
+	struct fid_cq *scq;
+	struct fid_cq *rcq;
+	struct peer_data data;
+};
+
+typedef struct application_data {
+	struct fid_mr *l_table_mr;
+	struct fid_mr *l_lock_mr;
+	struct fid_mr *l_scratch_mr;
+} app_data_t;
+
+typedef struct fabric {
+	struct fi_info *hints;
+	struct fi_info *info;
+	struct fid_fabric *fab;
+	struct fid_domain *dom;
+	struct fid_av *av;
+	void *addrs;
+	fi_addr_t *fi_addrs;
+	app_data_t data;
+	struct peer *peers;
+} fabric_t;
+
+fabric_t prov;
+
+int rank, num_ranks;
+uint64_t saved_initial_ran_val;
+
+struct test_tunables {
+	uint64_t total_mem_opt;
+	int num_updates_opt;
+};
+
+static struct test_tunables tunables = {0};
+
+static double usec_time(void)
+{
+	struct timeval tp;
+	gettimeofday(&tp, NULL);
+	return tp.tv_sec + tp.tv_usec / (double)1.0e6;
+}
+
+static inline void exchange_registration_data(fabric_t *fabric)
+{
+	uint64_t *keys;
+	void **addresses;
+	uint64_t local_table_key, local_lock_key;
+	int i;
+	app_data_t *data = &fabric->data;
+	struct peer *peers = fabric->peers;
+
+	keys = calloc(num_ranks, sizeof(*keys));
+	assert(keys);
+
+	addresses = calloc(num_ranks, sizeof(*addresses));
+	assert(addresses);
+
```

Comment:
First-touch in terms of which NUMA domain the physical memory is actually mapped to.  So if you allocate memory and free it here, depending on the implementation of malloc, it could be a subsequent allocation get this piece of memory expecting to have it automatically mapped to another NUMA domain.  That said, you can change the mapping with a system call. and people are addressing the sort of ad hoc manner in which this is done in various programming models.  Again, not relevant for this, but you asked.